### PR TITLE
Improvement/sso

### PR DESCRIPTION
**Pull request by jbpenrath**
## Purpose

Since we use sso on Richie through edx oauth2 provider, some mechanisms have to be improved:
- Sign up link targets django admin login page. Instead it should redirect on OAuth2 provider registration page.
- When the user logs out on Richie, he is still authenticated on LMS_BACKENDS. A logout on Richie should trigger log out on LMS_BACKENDS.
- If the user is already authenticated on OAuth2 provider, it should be also authenticated on Richie.

## Proposal

- [x] Add a registration oauth2 endpoint to be able to customize sign-up url.

- [x] If user is not authenticated on Richie, check if he is authenticated from OAuth2 provider. If yes, call automatically OAuth2 authorize endpoint.

- [x] Create a logout view which logs out user from all LMS_BACKENDS linked to richie.

### CODE REVIEW

File: src/richie/apps/core/context_processors.py [hunk 1]
```diff
@@ -32,6 +33,12 @@ def site_metas(request):
             "domain": site_current.domain,
             "web_url": f"{protocol:s}://{site_current.domain:s}",
         },
+        "SOCIAL": {
+            "registration_endpoint": getattr(
+                settings, "SOCIAL_AUTH_EDX_OAUTH2_REGISTRATION_ENDPOINT", ""
+            )
+            or reverse("admin:index")
```
Comment (slandrault) [inline src/richie/apps/core/context_processors.py:40]: Last parameter of `getattr` is the default value, so maybe write it like
```python
getattr(
    settings, "SOCIAL_AUTH_EDX_OAUTH2_REGISTRATION_ENDPOINT", reverse("admin:index")
)
```
to remove the `or` statement and simplify your code?

Comment (jbpenrath) [reply src/richie/apps/core/context_processors.py:40]: This syntax is less elegant but it is more resilient. In fact, getattr return default value only if the property does not exist. In the case of this settings exists, it do not return the default value.

By using `or reverse(...)`, we prevent this case.

File: src/frontend/js/data/useSession/index.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
```
Comment (lunika) [inline src/frontend/js/data/useSession/index.tsx:1]: You should add a module comment explaining what is this hook, why we need it, what you are doing, etc.

File: src/frontend/js/utils/test/factories.ts [hunk 3]
```diff
@@ -29,3 +30,27 @@ export const UserFactory = createSpec({
   full_name: faker.fake('{{name.firstName}} {{name.lastName}}'),
   username: faker.internet.email(),
 });
+
+interface CommonDataProps {
```
Comment (lunika) [inline src/frontend/js/utils/test/factories.ts:34]: Why not using the interface from `src/frontend/js/types/commonDataProps.ts` ?

File: src/frontend/js/types/commonDataProps.ts [hunk 4]
```diff
@@ -6,8 +6,10 @@ import { Nullable } from 'utils/types';
  */
 export interface CommonDataProps {
   context: {
+    allowed_hosts: string[];
```
Comment (lunika) [inline src/frontend/js/types/commonDataProps.ts:9]: I don't find where you are using it

File: src/richie/apps/core/context_processors.py [hunk 5]
```diff
@@ -32,13 +33,23 @@ def site_metas(request):
...
             {
                 "context": {
                     "csrftoken": get_token(request),
                     "environment": getattr(settings, "ENVIRONMENT", ""),
                     "release": getattr(settings, "RELEASE", ""),
                     "sentry_dsn": getattr(settings, "SENTRY_DSN", ""),
+                    "oauth2_whoami": getattr(
+                        settings, "SOCIAL_AUTH_EDX_OAUTH2_WHOAMI_ENDPOINT", None
+                    ),
+                    "allowed_hosts": getattr(settings, "ALLOWED_HOSTS", []),
```
Comment (lunika) [inline src/richie/apps/core/context_processors.py:52]: I don't know if it's something safe to expose this setting.

File: src/frontend/js/utils/isSafeURI.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import { location } from 'utils/indirection/window';
+
+const REDIRECT_WHITELIST =
+  (window as any).__richie_frontend_context__?.context?.redirect_whitelist || [];
```
Comment (lunika) [inline src/frontend/js/utils/isSafeURI.ts:4]: I don't find where you add this property in the context

File: docker-compose.yml [hunk 7]
```diff
@@ -1,10 +1,10 @@
-version: "3.5"
+version: '3.5'
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply docker-compose.yml:1]: My Prettier is a little bit intrusive 🤔

File: src/frontend/jest.config.js [hunk 8]
```diff
@@ -10,4 +10,5 @@ module.exports = {
   testMatch: [`${__dirname}/js/**/*.spec.+(ts|tsx|js)`],
   testURL: 'https://localhost',
   transformIgnorePatterns: ['node_modules/(?!(lodash-es)/)'],
+  coverageDirectory: '.coverage',
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/jest.config.js:13]: This directory is already ignored by git. So instead to add a new directory in `.gitignore`, I preferred to update the location to store jest coverage in an already ignored directory.
```
# Unit test / coverage reports
htmlcov/
.coverage
.cache
.pytest_cache
nosetests.xml
coverage.xml
```

File: src/frontend/js/components/UserLogin/index.tsx [hunk 9]
```diff
@@ -35,41 +35,54 @@ interface UserLoginProps {
   loginUrl: string;
   logoutUrl: string;
   signupUrl: string;
+  context: { [key: string]: any };
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/components/UserLogin/index.tsx]: You agree, I pass `oauth2_whoami` directly in props.

File: src/frontend/js/types/commonDataProps.ts (hunk 4 above)
Comment (jbpenrath) [reply src/frontend/js/types/commonDataProps.ts:9]: It's a mistake. I forgot to rename `allowed_hosts` into `redirect_whitelist`

File: src/richie/apps/core/context_processors.py (hunk 5 above)
Comment (jbpenrath) [reply src/richie/apps/core/context_processors.py]: As mentionned above, I will remove `oauth2_whoami` endpoint from the context to pass it directly in props.

File: src/frontend/js/utils/isSafeURI.ts (hunk 6 above)
Comment (jbpenrath) [reply src/frontend/js/utils/isSafeURI.ts:4]: Related to the mistake in `src/richie/apps/core/context_processors.py`

**Pull request closed without merging**